uaclient/gpg.py

```diff
@@ -0,0 +1,33 @@
+import logging
+import os
+from uaclient import exceptions
+from uaclient import util
+
+
+def export_gpg_key_from_keyring(
+        key_id: str, source_keyring_file: str,
+        destination_keyfile: str) -> None:
+    """Export a specific key from source_keyring_file into destination_keyfile
+
+    :param key_id: Long fingerprint of key to export.
+    :param source_keyring_file: The keyring file from which to export.
+    :param destination_keyfile: The filename created with the single exported
+        key.
+
+    :raise UserFacingError: Any GPG errors or if specific key does not exist in
+        the source_keyring_file.
+    """
+    export_cmd = ['gpg', '--output', destination_keyfile, '--yes',
+                  '--no-auto-check-trustdb', '--no-default-keyring',
+                  '--keyring', source_keyring_file, '--export', key_id]
+    logging.debug('Exporting gpg key %s from %s', key_id, source_keyring_file)
+    try:
+        util.subp(export_cmd)
+    except util.ProcessExecutionError as exc:
+        with util.disable_log_to_console():
+            logging.error(str(exc))
+        raise exceptions.UserFacingError(
+            "Unable to export gpg keys from keyring %s" % source_keyring_file)
+    if not os.path.exists(destination_keyfile):
+        raise exceptions.UserFacingError("gpg key '%s' not found in %s" % (
```

Comment:
`gpg` on trusty and eoan appears to output "`gpg: WARNING: nothing exported`" to stderr when we're in this error case; is it worth checking if that was output and emitting a more general message otherwise?

(I'm concerned there may be a class of "`gpg: WARNING: failed to frobnicate the doodad`" non-zero-exit error messages that we don't know about; if we treat them all the same, it could be quite frustrating for the user.)

What do you think?

Reply:
Added a check for 'nothing exported' in stderr. Will raise anything else if we didn't create the gpg keyring file properly